### PR TITLE
TaskTemplate offloading and available to task

### PR DESCRIPTION
**Pull request by kumare3**
Signed-off-by: Ketan Umare <ketan.umare@gmail.com>

# TL;DR
this PR updates the TaskReader interface to include a storage.DataReference (path) where the task template can be loaded from by the Container. It also creates helper implementation that lazily uploads the TaskTemplate to a protobuf store at the given reference when it is accessed and does it only once.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Tracking Issue
https://github.com/flyteorg/flyte/issues/850

### CODE REVIEW

File: go/tasks/plugins/array/awsbatch/transformer.go [hunk 1]
```diff
@@ -50,13 +50,26 @@ func FlyteTaskToBatchInput(ctx context.Context, tCtx pluginCore.TaskExecutionCon
 		return nil, errors.Errorf(errors.BadTaskSpecification, "config[%v] is missing", DynamicTaskQueueKey)
 	}
 
-	cmd, err := template.ReplaceTemplateCommandArgs(ctx, tCtx.TaskExecutionMetadata(), taskTemplate.GetContainer().GetCommand(), tCtx.InputReader(), tCtx.OutputWriter())
+	inputReader := array.GetInputReader(tCtx, taskTemplate)
```
Comment (wild-endeavor) [inline go/tasks/plugins/array/awsbatch/transformer.go:53]: is the inputReader the same?  previously it was `tCtx.InputReader()`

**Review by EngHabu — APPROVED** [review]
Would it be possible to try to either implement or at least go through the mental exercise of writing other plugins based on this? My worry is that exposing _just_ the task template feels like it'll be a short-sight as we try to enrich the FlyteKit-based plugins' writing experience and we will soon want to pass another Blah proto to the pod... Unfortunately, I do not know what that will contain just yet... just a gut feeling..

I would rather have had this implemented as "--task-plugin-data-path" param with a proto that -maybe for today- only has the task template... but tomorrow can add other things (maybe information about available resources or how to allocate resources... etc.)

I'm approving the PR though. Looks good for what it's trying to do.

File: go/tasks/pluginmachinery/ioutils/task_reader.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+	"github.com/flyteorg/flyteidl/gen/pb-go/flyteidl/core"
+
+	pluginsCore "github.com/flyteorg/flyteplugins/go/tasks/pluginmachinery/core"
+	"github.com/flyteorg/flytestdlib/atomic"
+	"github.com/flyteorg/flytestdlib/storage"
+	"github.com/pkg/errors"
+)
+
+var (
```
[the remark this answers is not in the document]
Comment (kumare3) [reply go/tasks/pluginmachinery/ioutils/task_reader.go:14]: just to make sure, we do not deviate from the core interface

File: go/tasks/pluginmachinery/core/template/template.go [hunk 3]
```diff
@@ -79,30 +85,28 @@ var outputRegex = regexp.MustCompile(`(?i){{\s*[\.$]OutputPrefix\s*}}`)
...
-	val = rawOutputDataPrefixRegex.ReplaceAllString(val, out.GetRawOutputPrefix().String())
+	val := inputFileRegex.ReplaceAllString(inputTemplate, params.Inputs.GetInputPath().String())
+	val = outputRegex.ReplaceAllString(val, params.OutputPath.GetOutputPrefixPath().String())
+	val = inputPrefixRegex.ReplaceAllString(val, params.Inputs.GetInputPrefixPath().String())
+	val = rawOutputDataPrefixRegex.ReplaceAllString(val, params.OutputPath.GetRawOutputPrefix().String())
 	val = perRetryUniqueKey.ReplaceAllString(val, perRetryKey)
 
-	inputs, err := in.Get(ctx)
+	// For Task template, we will replace only if there is a match. This is because, task template replacement
+	// may be expensive, as we may offload
```
[the remark this answers is not in the document]
Comment (kumare3) [reply go/tasks/pluginmachinery/core/template/template.go:99]: no, we offload only when ttt params.Task.Path is invoked! which is only if the taskTemplateRegex match happens. 
Again this is done so that its transparent to the templating method?
